### PR TITLE
feat: add two example apps that use nginx php8.2 and the sdk proxy

### PR DESCRIPTION
**Pull request by dakotawashok**
# Enhancement: New PHP 8.2 Hello World Examples with Nginx, Docker-Compose, and DevCycle Proxy

This Pull Request introduces two new Hello World applications for our PHP SDK, setup with PHP 8.2 and integrated with nginx, docker-compose, and the DevCycle Local Bucketing Proxy.

## Key Changes:

1. **New Hello World Applications:** Two examples added to demonstrate the PHP SDK's functionality with the DevCycle Local Bucketing Proxy in a PHP 8.2 environment.
2. **PHP 8.2 Compatibility:** These containers assume PHP 8.2 is being used
3. **Nginx Integration:** Nginx is used as the web server
4. **Docker-Compose Setup:** The environment setup is streamlined with docker-compose for ease of use and consistency.
5. **DevCycle Local Bucketing Proxy Integration:** 

Looking forward to your feedback on these enhancements.

Dakota Washok
Senior Software Engineer/DevOps Engineer

### CODE REVIEW

**Comment by dakotawashok** [comment]
Looks like there are some failing tests reported by the Harness Tests suite? I'm not sure if those are okay or not, but let me know if you want me to update anything!

**Comment by jonathannorris** [comment]
Hey @dakotawashok thanks so much for submitting this PR! Our "virtual office" is closed this week, we will be sure to dig into it next week when everyone is back.

**Comment by JamieSinn** [comment]
Further investigation shows this is due to a file permissions issue for the raw socket file:
https://github.com/DevCycleHQ/sdk-proxy/issues/12#issuecomment-1874438915

This is an awesome contribution for the repo!

**Comment by dakotawashok** [comment]
Hello @JamieSinn / @jonathannorris !
I'm not sure who to ask this, so I figured I'd ping both, sorry about that.
We decided to use the HTTP SDK proxy instead of the UDP proxy for the time being, until we can further investigate our setup and how to integrate the proper file/group permissions across our services. 
The HTTP SDK Proxy seems to be working like a charm, though! Thanks for that :)

As for this PR, though, I still think it'd be good to have a variety of example setups in this project for others to use and build off of. If yall agree, I'd love to get this in here still. Is there anything you need me to update on my end?

**Comment by JamieSinn** [comment]
Absolutely, I'd love to get it in here.

I think for tweaks, making the socket available under 777 would be a quick fix, somehow part of the docker compose setup or a docker file edit.

With that, it works and we can merge the rest as is!

**Comment by dakotawashok** [comment]
Hey @JamieSinn I've altered the Supervisord configuration in the UDP example app to alter the socket file permissions after startup. Please take a look and let me know if you want anything else updated!

File: examples/hello-world-udp-proxy/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,26 @@
+FROM php:8.2-fpm
+
+ENV DEVCYCLE_SDK_KEY="xxx"
```
Comment (JamieSinn) [inline examples/hello-world-udp-proxy/Dockerfile:3]: Docker arg for these? it's in the compose file already

File: examples/hello-world-udp-proxy/supervisord.conf [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+[supervisorctl]
+
+[program:nginx]
+command=/usr/sbin/nginx -g 'daemon off;'
+
+[program:php-fpm]
+command=/usr/local/sbin/php-fpm -F
+
+[program:dvc-lb-proxy]
+command=/bin/bash -c "/usr/bin/devcycle-sdk-proxy -config /etc/devcycle-sdk-proxy/dvc-proxy-config.json && sudo chmod 777 /tmp/dvc_lb_proxy.sock"
```
Comment (JamieSinn) [inline examples/hello-world-udp-proxy/supervisord.conf:13]: Might want to add a sleep in between - unless you tested this and it starts up fine

Comment (dakotawashok) [reply examples/hello-world-udp-proxy/supervisord.conf:13]: I have not tested this update. Let me test it out real quick and determine if the sleep is necessary. Good catch!

Comment (dakotawashok) [reply examples/hello-world-udp-proxy/supervisord.conf:13]: @JamieSinn Even with the sleep in between these commands, the file permissions aren't being set by the chmod command. I tried adding a `ENTRYPOINT` script with the command running after the supervisord command start, but that didn't work either. I'm still a bit green with creating Dockerfiles and docker-compose files, do you have any suggestions on what I might be doing wrong?

Comment (JamieSinn) [reply examples/hello-world-udp-proxy/supervisord.conf:13]: Generally the tricky part is you're calling a single command shell, the easiest solution might be to create a script that handles the delay logic for you and then call that.

Alternatively, I could just merge the pr that's making it natively accessible in the proxy (the permissions)

**Comment by JamieSinn** [comment]
Hey @dakotawashok - are you able to update the example apps to use the new init flow?
Check the readme if you need help

**Comment by dakotawashok** [comment]
@JamieSinn I'll take a look and integrate it with these example apps once I'm able to hop back onto this, sorry for the wait!

**Comment by JamieSinn** [comment]
No rush! Thank you!

**Comment by dakotawashok** [comment]
I've got the issues you've pointed out resolved, could you take another look at this @JamieSinn ? I'll test again in the morning, but I think these changes are good to go! Let me know if you'd like me to resolve the unit tests that are failing, as I can't remember if they were a part of the Checks in this repo when I originally opened this haha

**Comment by JamieSinn** [comment]
Looks great, I'll take a shot at testing it when I'm in the office

**Comment by JamieSinn** [comment]
Seems like supervisord is crashing still 

image
Did it work when you tried it locally? I'm using a test token so it may be a weird environment setup

File: examples/hello-world-udp-proxy/app.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+require_once(__DIR__ . '/vendor/autoload.php');
+
+use DevCycle\DevCycleConfiguration;
+use DevCycle\Api\DevCycleClient;
+use DevCycle\Model\DevCycleOptions;
+use DevCycle\Model\DevCycleUser;
+
+$options = new DevCycleOptions(
+    enableEdgeDB: false,
+    bucketingApiHostname: "http://localhost",
```
Comment (JamieSinn) [inline examples/hello-world-udp-proxy/app.php:12]:
```suggestion
    bucketingApiHostname: "http:/localhost",
```

Only one slash for socket paths

File: examples/hello-world-udp-proxy/app.php [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+);
+
+$apiInstance = new DevCycleClient(
+    getenv("DEVCYCLE_SDK_KEY"),
+    dvcOptions: $options
+);
+
+$user_data = new DevCycleUser(array(
+    'user_id' => 'integration-test',
+    'email' => 'nhaynes@adaction.com',
```
Comment (JamieSinn) [inline examples/hello-world-udp-proxy/app.php:23]:
```suggestion
    'email' => 'email@example.com',
```

File: examples/hello-world-http-proxy/app.php [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+);
+
+$apiInstance = new DevCycleClient(
+    getenv("DEVCYCLE_SDK_KEY"),
+    dvcOptions: $options
+);
+
+$user_data = new DevCycleUser(array(
+    'user_id' => 'integration-test',
+    'email' => 'nhaynes@adaction.com',
```
Comment (JamieSinn) [inline examples/hello-world-http-proxy/app.php:22]:
```suggestion
    'email' => 'email@example.com',
```

File: examples/hello-world-udp-proxy/app.php (hunk 3 above)
Comment (dakotawashok) [reply examples/hello-world-udp-proxy/app.php:12]: I've never seen these commit suggestions before, I'll have to make use of them for our company PRs! I'll resolve this now, though.

Comment (JamieSinn) [reply examples/hello-world-udp-proxy/app.php:12]: They're super handy for one-line changes - would immensely recommend.

**Comment by dakotawashok** [comment]
Looks good now! Here's the result from variable.php for a variable set to true in the http proxy example: 
image
Here's the result from variable.php for a variable set to true in the udp proxy example: 
image

**Comment by dakotawashok** [comment]
@JamieSinn can you please try this again, replacing the sdk in the docker-compose files with a proper key? This should be ready for review.

File: examples/hello-world-udp-proxy/dvc-proxy-config.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+  "instances": [
+    {
+      "sdkKey": "${DEVCYCLE_SDK_KEY}",
+      "unixSocketPath": "/tmp/dvc_lb_proxy.sock",
+      "unixSocketEnabled": true,
```
Comment (JamieSinn) [inline examples/hello-world-udp-proxy/dvc-proxy-config.json:6]:
```suggestion
      "unixSocketEnabled": true,
      "unixSocketPermissions": 777,
```

You'll need to set the file perms here too.

Comment (dakotawashok) [reply examples/hello-world-udp-proxy/dvc-proxy-config.json:6]: I've removed the extra commands from the supervisord configuration and I'm testing this functionality now

Comment (dakotawashok) [reply examples/hello-world-udp-proxy/dvc-proxy-config.json:6]: @JamieSinn Is this configuration option supposed to negate the need to alter the socket file permissions after the proxy command is run?

Comment (JamieSinn) [reply examples/hello-world-udp-proxy/dvc-proxy-config.json:6]: yep!

Comment (dakotawashok) [reply examples/hello-world-udp-proxy/dvc-proxy-config.json:6]: I'm not able to get the actual variable value when using this configuration option and removing the extra commands for altering the socket file permissions, which is why I ask.

Comment (JamieSinn) [reply examples/hello-world-udp-proxy/dvc-proxy-config.json:6]: Hmm - that's odd - let me do a quick test.

Comment (dakotawashok) [reply examples/hello-world-udp-proxy/dvc-proxy-config.json:6]: I'll push up the changes I used to test that, I think that'd help, and I can resolve any issues afterward.

Comment (JamieSinn) [reply examples/hello-world-udp-proxy/dvc-proxy-config.json:6]: sounds good. I have a feeling this might be on my implementation.

Comment (JamieSinn) [reply examples/hello-world-udp-proxy/dvc-proxy-config.json:6]: Yep - looks like that was on me. I fixed it to now support octal perms. 1.1.1 will fix it

**Comment by dakotawashok** [comment]
@JamieSinn I'm still getting an issue, even with updating the proxy package to 1.1.1. Here's the error I get when running the command now, a new error: 
```
2024/02/13 21:08:17 Loading configuration from file: /etc/devcycle-sdk-proxy/dvc-proxy-config.json
2024/02/13 21:08:17 Failed to parse config: failed to parse config from JSON: json: cannot unmarshal number into Go struct field ProxyInstance.instances.unixSocketPermissions of type string
2024/02/13 21:08:17 Please either set the config path or set the environment variables
```

I tried `"unixSocketPermissions": "777",` and `"unixSocketPermissions": 777,` in my `dvc-proxy-config.json`.

Here's the results of the udp variable test. Interestingly, the webpage still is loading without a devcycle error being reported, even though the proxy isn't apparently running:
image

**Comment by JamieSinn** [comment]
> @JamieSinn I'm still getting an issue, even with updating the proxy package to 1.1.1. Here's the error I get when running the command now, a new error:
> 
> ```
> 2024/02/13 21:08:17 Loading configuration from file: /etc/devcycle-sdk-proxy/dvc-proxy-config.json
> 2024/02/13 21:08:17 Failed to parse config: failed to parse config from JSON: json: cannot unmarshal number into Go struct field ProxyInstance.instances.unixSocketPermissions of type string
> 2024/02/13 21:08:17 Please either set the config path or set the environment variables
> ```
> 
> I tried `"unixSocketPermissions": "777",` and `"unixSocketPermissions": 777,` in my `dvc-proxy-config.json`.
> 
> Here's the results of the udp variable test. Interestingly, the webpage still is loading without a devcycle error being reported, even though the proxy isn't apparently running: image
ah my bad - I forgot to say you'll need to change that param to be a string, `"0777"` specifically.

File: examples/hello-world-udp-proxy/dvc-proxy-config.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+      "unixSocketPermissions": 777,
```
Comment (JamieSinn) [inline examples/hello-world-udp-proxy/dvc-proxy-config.json:7]:
```suggestion
      "unixSocketPermissions": "0777",
```

**Comment by dakotawashok** [comment]
Woohoo! That worked! I'm able to get our variable successfully now on both proxy hello world examples. If this is good to go for yall, please feel free to merge in.

**Review by JamieSinn — APPROVED** [review]

File: examples/hello-world-udp-proxy/dvc-proxy-config.json (hunk 7 above)
Comment (dakotawashok) [reply examples/hello-world-udp-proxy/dvc-proxy-config.json]: I forgot octal permissions had were a specific thing with the "0" in front 🤦🏻

Comment (JamieSinn) [reply examples/hello-world-udp-proxy/dvc-proxy-config.json]: yep - gotta love filemodes!

**Pull request merged**